Juhitha took over for Jiayu & Shuhua - All Inventory Types page functionalities

Shuhua-L opened the pull request:
# Description

Implement the functionalities of the All Inventory Types page, which the UI was implemented in this [PR](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1868)

## Related PRS (if any):

To test this backend PR you need to checkout the [#3833](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/3833) frontend PR.

## Main changes explained:

- Added missing PUT routes for updating inventory types (Reusables, Equipments, Consumables, Tools)
- Added missing DELETE routes for removing inventory types across all categories
- Fixed POST route for Reusables inventory type (was completely missing)
- Added missing POST/DELETE routes for inventory units management
- Updated bmInventoryTypeRouter.js with comprehensive CRUD operations for all inventory types
- Integrated controller methods for updateSingleInvType, deleteSingleInvType, addInvUnit, deleteInvUnit

## How to test:

1. Check into current branch Shuhua-Types_List_Page_Edit_Delete
2. Do npm install && npm start to run this PR locally on port 4500
3. Checkout the frontend [cleanup branch](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/3833)
4. Verify all inventory type operations work:
  - GET: Fetch Materials, Consumables, Tools, Equipment, Reusables
  - POST: Add new inventory types for all categories
  - PUT: Update existing inventory types
  - DELETE: Remove inventory types

Shuhua-L committed:
added routers, controllers for adding/deleting units

Shuhua-L committed:
added routers, controllers for adding/deleting units

Shuhua-L committed:
added routers updating/deleting types

Shuhua-L committed:
added routers updating/deleting types

Shuhua-L committed:
added routers, controllers for adding/deleting types

Shuhua-L committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

Shuhua-L committed:
Merge branch 'Shuhua-Types_List_Page_Edit_Delete' of https://github.com/OneCommunityGlobal/HGNRest into Shuhua-Types_List_Page_Edit_Delete

Shuhua-L committed:
fix plural of equipment for consistency

Shuhua-L committed:
change all categories to singular for consistency

skaush21 commented:
left comment on frontend [PR #1984](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-1899574683)

brandta-1 reviewed (changes requested):
# Review
Refer to [front end review](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-1899586867).

skaush21 reviewed (approved):
comment on frontend [PR 1984](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-1899574683)

XiaohanMeng reviewed (approved):
Hi, I left my comment on FE project: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-1904912733

ShengjieMao reviewed (approved):
Left review in frontend pr

20chen-7 reviewed (commented):
Good job! The add/delete works as described. 
<img width="1263" alt="Screenshot 2024-03-15 at 8 54 18 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/6e4e9797-ddb5-4684-b426-cfdb4ab61be8">

<img width="1252" alt="Screenshot 2024-03-15 at 8 55 14 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/10b0429e-84e4-44db-84ae-d93462df9324">
<img width="1245" alt="Screenshot 2024-03-15 at 8 56 03 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/f2b5831f-bd8b-4dcb-899f-2438bf373f6a">

<img width="1267" alt="Screenshot 2024-03-15 at 8 56 13 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/f53fa320-e8fa-4336-a44f-2ec0b4299380">

But I strongly recommend add back to previous btn here.

Everything is wonderful!
<img width="1440" alt="Screenshot 2024-03-15 at 8 56 40 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/3349f0d8-4438-4d71-9ee3-8e286a7cab94">
<img width="1051" alt="Screenshot 2024-03-15 at 8 56 46 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/3f5b8747-4243-4229-b12e-a6a2d174ef22">
<img width="1174" alt="Screenshot 2024-03-15 at 8 57 10 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/a72b1195-371f-414c-bf75-73ff7adc0716">
…]()
<img width="1137" alt="Screenshot 2024-03-15 at 8 57 38 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/b978e42e-2b10-4cfd-9c01-d9d509d4c8ae">
<img width="1144" alt="Screenshot 2024-03-15 at 8 57 43 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/102696148/20503131-7706-41ea-b094-f42a05319ebc">

yvonnep165 reviewed (approved):
Hi @Shuhua-L , I tested this PR with the frontend PR 1984 and everything worked as intended.  I left a comment in the frontend PR.

peter6866 committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

peter6866 committed:
fixed bug

peter6866 committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

kabraambika reviewed (approved):
I have provided comment in frontend PR #1984

varunreddy03 reviewed (approved):
I have provided my comments in the related frontend PR #1984.

adityasure9 reviewed (changes requested):
Hi @Shuhua-L,

I tested the PR according to the instructions, and I encountered the following issues:

After editing and saving changes on the Inventory Types page, the entire page becomes unresponsive and gets stuck. This issue prevents further interaction with the page, making it difficult to proceed with other actions like adding or deleting items.
I noticed similar errors as mentioned by other reviewers, particularly related to the frontend not loading properly due to compile-time errors.
Given these issues, it seems there are some bugs that need to be addressed before this can be merged. Please let me know if you need any further details or testing on my end.

https://github.com/user-attachments/assets/64a55d0f-2e22-4cf1-8ac2-b54143663603

Thanks for your efforts on this!

SMitta11 reviewed (changes requested):
Added review in frontend PR 1984

Shreyav2000 reviewed (commented):
Hi, please refer to the below link for the review I left on the Frontend PR:
https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-2287439514

Parth-tech commented:
Hey Jiayu,
I have left a review on the frontend repo. [link](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-2304398820)

sud306 commented:
Hello,
I tried to test PR1984 with PR765 but it gave error on frontend.

Muzammil603 reviewed (commented):
added review on pr #1984 front end

SFA23SCM35V reviewed (changes requested):
A blank screen is displayed when attempting to access the application.
![1](https://github.com/user-attachments/assets/8784d40e-3a65-4ede-b5fb-c135b2de32c0)
![2](https://github.com/user-attachments/assets/0069817d-b29d-4842-a308-91e8e0c2b643)
![3](https://github.com/user-attachments/assets/8c8aaa3b-73ce-48a3-9868-4d3f464b5799)

neeharikakoniki reviewed (changes requested):
Reviwed at https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-2330304075

mrinalini-raghavendran19 reviewed (approved):
Provided feedback on related frontend PR!

sheetalmangate commented on src/controllers/bmdashboard/bmInventoryTypeController.js:
```diff
@@ -363,35 +362,110 @@ function bmInventoryTypeController(InvType, MatType, ConsType, ReusType, ToolTyp
+      );
       if (!updatedInvType) {
-        return res.status(404).json({ error: 'invType Material not found check Id' });
+        res.status(404).json({ error: 'invTypeId does not exist' });
```
It would be better to use general error message instead of using variable name "invTypeId" in error message.

zhiminV reviewed (changes requested):
I have faced error when open the frontend:https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-2397389542 

![Screenshot 2024-10-26 at 7 54 26 PM](https://github.com/user-attachments/assets/55150590-28e2-4efa-babd-876c65cef244)

humera314 commented:
left comment on https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984

Supriya1659 reviewed (approved):
I have added my comment on the frontend PR #1984
There is no changes in the backend but only the file name changes in the frontend.

honglinchen0524 commented:
Hi! I've added my comment on the frontend PR #1984, great work!

jeyanthi-sm commented:
I have added my comment to the frontend PR - [https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#issuecomment-2608262046](Link)

SSP24SCM16G reviewed (commented):
Please refer to frontend review https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1984#pullrequestreview-2635415089 

![Uploading Screenshot 2025-02-22 at 22.44.20.png…]()

rohith1667 reviewed (approved):
Reviewed frontend pr 1984 all the changes are implemented. all the features are working as mentioned.


https://github.com/user-attachments/assets/56116064-c23f-4afa-9dbd-0d763fd829ff

HSP24SCM69K reviewed (approved):
Reviewed Frontend PR #1984 – All requested changes have been successfully implemented, and all features are functioning as expected.
<img width="1470" alt="Screenshot 2025-03-08 at 13 37 55" src="https://github.com/user-attachments/assets/fd520cbf-033d-4dcb-828f-962bd0d09f07" />
<img width="1470" alt="Screenshot 2025-03-08 at 13 38 05" src="https://github.com/user-attachments/assets/431df69e-2dfb-41dd-b473-ae07a2ef5e72" />
<img width="1470" alt="Screenshot 2025-03-08 at 13 38 45" src="https://github.com/user-attachments/assets/f41b226e-e51d-4465-86f0-4680dab295e8" />
<img width="1470" alt="Screenshot 2025-03-08 at 13 38 48" src="https://github.com/user-attachments/assets/98697212-961f-4085-9d8b-9600b850b3c4" />
<img width="1470" alt="Screenshot 2025-03-08 at 13 39 04" src="https://github.com/user-attachments/assets/24d63523-d41a-4bc4-a6d2-bc01480490e1" />
![Screenshot 2025-03-08 at 13 39 27](https://github.com/user-attachments/assets/87deb065-a0e0-49f8-ab2a-14a5388a5c19)
<img width="1470" alt="Screenshot 2025-03-08 at 13 39 33" src="https://github.com/user-attachments/assets/de247497-3283-4603-a63c-a5b4cbbc0480" />

rishwa344 reviewed (changes requested):
I reviewed this PR and verified that adding, editing, and deleting inventory types are working as expected. However, after editing an inventory type, the page becomes unresponsive and throws a console error. The issue resolves only after refreshing the page. Other functionalities, including adding and deleting inventory units, are working fine.
<img width="1433" alt="Screenshot 2025-03-10 at 1 18 41 PM" src="https://github.com/user-attachments/assets/b7387ac3-10a0-446d-8e45-dbd388803812" />
<img width="1303" alt="Screenshot 2025-03-10 at 1 19 33 PM" src="https://github.com/user-attachments/assets/1f4eafaa-a795-491d-9209-5994c23e4e40" />

sriramsme committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

sri0606 commented:
@everyone please re-review this PR. Updates were made on frontend.

humemah commented:
Followed the steps and verified that the API handles inventory updates, unit management, responses, and error messages correctly.
![image](https://github.com/user-attachments/assets/1d8c149a-700d-462d-8798-1f833c11503e)
![image](https://github.com/user-attachments/assets/9a69c1b4-7024-4809-9c17-a9766b766b33)
![image](https://github.com/user-attachments/assets/f2e19f1b-d010-4d1b-b509-93926a0f6288)

Commit:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

Juhitha-Reddy committed:
Updated Add and Edit Inventory APIs

Juhitha-Reddy committed:
Fixed controller routes

Juhitha-Reddy committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

Anusha-Gali reviewed (approved):
Hi Juhitha,

I have reviewed your PR locally and have mentioned my comments in the frontend PR: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/3833#pullrequestreview-3611721651

Juhitha-Reddy committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

Juhitha-Reddy committed:
updated the delete api for typelist

Juhitha-Reddy committed:
fixed text issue

Juhitha-Reddy committed:
Fix inventory types CRUD: add routes, discriminator model resolution, and reusable type endpoint

- Add /invtypes/:type/:invtypeId route for PUT and DELETE to match frontend URL
- Fix Mongoose discriminator issue: resolve correct subtype model (MatType,
  ConsType, etc.) in updateInventoryType and updateNameAndUnit so fields like
  unit and fuelType are no longer silently dropped by the base model
- Add addReusableType controller and POST /invtypes/reusables route
- Add addInventoryUnit and deleteInventoryUnit controller functions and routes

Made-with: Cursor

Juhitha-Reddy committed:
Merge branch 'development' into Shuhua-Types_List_Page_Edit_Delete

one-community commented:
Thank you all, merging!
